src/Web/Slack.hs

```diff
@@ -564,24 +592,68 @@ historyFetchAll
-    -- In most cases an application will want the most recent messages
-    -- and will page backward from there.
-    --
-    -- for reference (does not apply here) => If oldest is provided but not
-    -- latest then the messages returned are those closest to oldest,
-    -- allowing you to page forward through history if desired.
-    rsp <- makeReq $ Common.HistoryReq channel count (Just latest) (Just oldest) False
-    case rsp of
-      Left _ -> return rsp
-      Right (Common.HistoryRsp msgs hasMore) -> do
-          let oldestReceived = Common.messageTs <$> lastZ msgs
-          if not hasMore || isNothing oldestReceived
-              then return rsp
-              else mergeResponses msgs <$>
-                   historyFetchAll makeReq channel count oldest (fromJust oldestReceived)
+historyFetchAll makeReq channel count =
+  commonHistoryFetchAll $ \latest oldest ->
+    makeReq . Common.HistoryReq channel count latest oldest
+
+
+-- | Returns an action to send a request to get the history of a conversation.
+--
+--   To fetch all messages in the conversation, run the returned 'LoadPage' action
+--   repeatedly until it returns an empty list.
+conversationsHistoryAll
+  :: (MonadReader env m, HasManager env, HasToken env, MonadIO m)
+  =>  Conversation.HistoryReq
+  -- ^ The first request to send. _NOTE_: 'Conversation.historyReqCursor' is silently ignored.
+  -> m (LoadPage m [Common.Message])
+  -- ^ An action which returns a new page of messages every time called.
+  --   If there are no pages anymore, it returns an empty list.
+conversationsHistoryAll = conversationsHistoryAllBy conversationsHistory
+
+
+-- | Returns an action to send a request to get the replies of a conversation.
+--
+--   To fetch all replies in the conversation, run the returned 'LoadPage' action
+--   repeatedly until it returns an empty list.
+--
+--   *NOTE*: The conversations.replies endpoint always returns the first message
+--           of the thread. So every page returned by the 'LoadPage' action includes
+--           the first message of the thread. You should drop it if you want to
+--           collect messages in a thread without duplicates.
+repliesFetchAll
+  :: (MonadReader env m, HasManager env, HasToken env, MonadIO m)
+  =>  Conversation.RepliesReq
+  -- ^ The first request to send. _NOTE_: 'Conversation.repliesReqCursor' is silently ignored.
+  -> m (LoadPage m [Common.Message])
+  -- ^ An action which returns a new page of messages every time called.
+  --   If there are no pages anymore, it returns an empty list.
+repliesFetchAll = repliesFetchAllBy conversationsReplies
+
+
+commonHistoryFetchAll
+  :: (MonadReader env m, HasManager env, HasToken env, MonadIO m)
+  => (Maybe Common.SlackTimestamp -> Maybe Common.SlackTimestamp -> Bool -> m (Response Common.HistoryRsp))
+  -> Common.SlackTimestamp
+  -> Common.SlackTimestamp
+  -> m (Response Common.HistoryRsp)
+commonHistoryFetchAll makeReq oldest latest = do
```

Comment:
Oops, this is actually used by the existing functions. I think we should deprecate the old paginators because both `mpimHistory` and `channelsHistory` will be unavailable due to the coming updates of Slack API.
But I'll do it later anyway.